sklearn/ensemble/_hist_gradient_boosting/tests/test_splitting.py

```diff
@@ -246,16 +256,109 @@ def test_min_gain_to_split():
     sum_hessians = all_hessians.sum()
     hessians_are_constant = False
 
-    actual_n_bins = np.array([n_bins] * X_binned.shape[1],
-                             dtype=np.uint32)
     builder = HistogramBuilder(X_binned, n_bins, all_gradients,
                                all_hessians, hessians_are_constant)
-    splitter = Splitter(X_binned, actual_n_bins,
+    n_bins_non_missing = np.array([n_bins - 1] * X_binned.shape[1],
+                                  dtype=np.uint32)
+    has_missing_values = np.array([False] * X_binned.shape[1], dtype=np.uint8)
+    missing_values_bin_idx = n_bins - 1
+    splitter = Splitter(X_binned, n_bins_non_missing, missing_values_bin_idx,
+                        has_missing_values, l2_regularization,
+                        min_hessian_to_split, min_samples_leaf,
+                        min_gain_to_split, hessians_are_constant)
+
+    histograms = builder.compute_histograms_brute(sample_indices)
+    split_info = splitter.find_node_split(n_samples, histograms,
+                                          sum_gradients, sum_hessians)
+    assert split_info.gain == -1
+
+
+@pytest.mark.parametrize(
+    'X_binned, all_gradients, has_missing_values, n_bins_non_missing, '
+    ' expected_bin_idx, expected_go_to_left', [
+
+        # basic sanity check with no missing values: given the gradient
+        # values, the split must occur on bin_idx=3
+        ([0, 1, 2, 3, 4, 5, 6, 7, 8, 9],  # X_binned
+         [1, 1, 1, 1, 5, 5, 5, 5, 5, 5],  # gradients
+         False,  # no missing values
+         10,  # n_bins_non_missing
+         3,  # expected_bin_idx
+         'not_applicable'),
+
+        # We replace 2 samples by NaNs (bin_idx=9)
+        # These 2 samples were mapped to the left node before, so they should
+        # be mapped to left node again
+        # Notice how the bin_idx threshold changes from 3 to 1.
+        ([9, 0, 1, 9, 2, 3, 4, 5, 6, 7],  # 9 <=> missing
+         [1, 1, 1, 1, 5, 5, 5, 5, 5, 5],
+         True,  # missing values
+         8,  # n_bins_non_missing
```

Comment:
shouldn't this be 9? Why is it 8?

Reply:
because max_bins is 9. The bin 8 exists, but it's just not here.

I kept 9 as max_bins for consistency with the previous cases.

Reply:
yeah exactly. 9 is supposed to be the bin for missing value, which means n_bins_non_missing is 9

Reply:
Sorry for the poor explanation.
You can actually set it to 9, the test would still pass. Since there is no sample at bin 8, the histogram for this bin is 0 everywhere, and it is thus implicitly "ignored".

Reply:
Note we only check for the values of `expected_bin_idx` and `expected_go_to_left`. The rest of the parametrized arguments are actually inputs to the test

Reply:
If you set it to 9, the test fails on the patch I sent to your branch

Reply:
Then maybe there's something wrong with the bounds when we scan the bins?

Reply:
actually I don't think that 9 would make sense.

That would mean that the binner discovers 9 bins for non-missing values, but yet only 8 of these bins are used? I don't think this can/should happen. The code runs fine even if you pass 9, but that's not a use-case.

Note that having `n_bins_non_missing` to 8 while `9` is the bin for missing values is a perfectly normal case: user set `max_bins` to 9, and the feature that we're considering only has 8 unique values.

Reply:
yeah fair point, forgot about that.